pkg/api/handlers/catalogs.go

```diff
@@ -0,0 +1,385 @@
+	}
+
+	if err := req.Create(&entry); err != nil {
+		return fmt.Errorf("failed to create entry: %w", err)
+	}
+
+	return req.Write(convertMCPServerCatalogEntry(entry))
+}
+
+func (h *MCPCatalogHandler) UpdateEntry(req api.Context) error {
+	catalogName := req.PathValue("catalog_id")
+	entryName := req.PathValue("entry_id")
+
+	var entry v1.MCPServerCatalogEntry
+	if err := req.Get(&entry, entryName); err != nil {
+		return fmt.Errorf("failed to get entry: %w", err)
+	}
+
+	if entry.Spec.MCPCatalogName != catalogName {
+		return fmt.Errorf("entry does not belong to catalog")
+	}
+
+	if !entry.Spec.Editable {
+		return fmt.Errorf("entry is not editable")
+	}
+
+	var manifest types.MCPServerCatalogEntryManifest
+	if err := req.Read(&manifest); err != nil {
+		return fmt.Errorf("failed to read entry manifest: %w", err)
+	}
+
+	hasCommand, hasURL, err := h.validateMCPServerCatalogEntryManifest(manifest)
+	if err != nil {
+		return fmt.Errorf("failed to validate entry manifest: %w", err)
+	}
+
+	if hasCommand {
+		entry.Spec.CommandManifest = manifest
+		entry.Spec.URLManifest = types.MCPServerCatalogEntryManifest{}
+	} else if hasURL {
+		entry.Spec.URLManifest = manifest
+		entry.Spec.CommandManifest = types.MCPServerCatalogEntryManifest{}
+	} else {
+		// Should be impossible since we validated this earlier.
+		return fmt.Errorf("invalid manifest")
+	}
+
+	if err := req.Update(&entry); err != nil {
+		return fmt.Errorf("failed to update entry: %w", err)
+	}
+
+	return req.Write(convertMCPServerCatalogEntry(entry))
+}
+
+func (h *MCPCatalogHandler) DeleteEntry(req api.Context) error {
+	catalogName := req.PathValue("catalog_id")
+	entryName := req.PathValue("entry_id")
+
+	var entry v1.MCPServerCatalogEntry
+	if err := req.Get(&entry, entryName); err != nil {
+		return fmt.Errorf("failed to get entry: %w", err)
+	}
+
+	if entry.Spec.MCPCatalogName != catalogName {
+		return fmt.Errorf("entry does not belong to catalog")
+	}
+
+	if !entry.Spec.Editable {
+		return fmt.Errorf("entry is not editable and cannot be manually deleted")
+	}
+
+	if err := req.Delete(&entry); err != nil {
+		return fmt.Errorf("failed to delete entry: %w", err)
+	}
+
+	return nil
+}
+
+func (h *MCPCatalogHandler) validateMCPServerCatalogEntryManifest(manifest types.MCPServerCatalogEntryManifest) (bool, bool, error) {
+	if manifest.Server.Name == "" {
+		return false, false, fmt.Errorf("server name is required")
+	}
+
+	var (
+		hasCommand, hasURL bool
+	)
+	if manifest.Server.Command != "" {
+		hasCommand = true
+
+		if manifest.Server.Command == "docker" {
+			if len(manifest.Server.Args) == 0 || len(h.allowedDockerImageRepos) > 0 && !slices.ContainsFunc(h.allowedDockerImageRepos, func(s string) bool {
+				return strings.HasPrefix(manifest.Server.Args[len(manifest.Server.Args)-1], s)
+			}) {
+				return false, false, fmt.Errorf("docker command must be followed by a valid image name from one of the allowed repos (%s)", strings.Join(h.allowedDockerImageRepos, ", "))
+			}
```

Comment:
TODO: probably worthwhile to validate to make sure the command is one of ["docker", "uvx", "npx"], etc.